### PR TITLE
Support Simple and Advanced Datums

### PR DESCRIPTION
**Pull request by keyan-m**
To allow further flexibility, this PR proposes the datum to be updated so that:
- The `owner` becomes optional, allowing a closed reclaim endpoint
- The routing fee is specified on a per UTxO basis instead of being hard-coded (minimum of 0)

### CODE REVIEW

**Review by mpetruska — APPROVED** [review]
:+1:, very nice job! Please update the documentation (README.md) as well.

**Comment by keyan-m** [comment]
As suggested by @solidsnakedev `SmartHandleDatum` is now a sum type, supporting either a `Simple` datum, where only an `owner` is stored, or an `Advanced` datum for more complex validator functions.

**Comment by solidsnakedev** [comment]
LGTM

File: src/StakingValidator.hs [hunk 1]
```diff
@@ -51,6 +54,7 @@ data RouterRedeemer = RouterRedeemer
 PlutusTx.makeLift ''RouterRedeemer
 PlutusTx.makeIsDataIndexed ''RouterRedeemer [('RouterRedeemer, 0)]
 
+-- TODO: Switch to: '[ "indices" ':= PBuiltinList (PAsData (PBuiltinPair (PAsData PInteger) (PAsData PInteger)))
```
Comment (mpetruska) [inline src/StakingValidator.hs:57]: Is this a missed thing, or for the future?

File: src/Utils.hs [hunk 2]
```diff
@@ -140,8 +153,44 @@ pfoldl2 =
...
+              ( \b bs ->
+                  pelimList
+                    (\c cs -> self # (func # acc # a # b # c) # as # bs # cs)
+                    perror
+                    lc
+              )
+              (pif (pnull # lc) acc perror)
+              lb
+        )
+        (pif (pnull # lb) acc perror)
```
Comment (mpetruska) [inline src/Utils.hs]:
```suggestion
        (pif (pand (pnull # lb) (pnull # lc)) acc perror)
```

File: src/StakingValidator.hs [hunk 3]
```diff
@@ -136,18 +169,61 @@ puniqueOrdered =
           )
      in go
 
-smartHandleStakeValidatorW :: Term s ((PAddress :--> PData :--> PDatum :--> PBool) :--> PAddress :--> PStakeValidator)
-smartHandleStakeValidatorW = phoistAcyclic $ plam $ \validateFn swapAddress redeemer ctx -> P.do
+smartHandleStakeValidatorW :: Term s (PCustomValidator :--> PAddress :--> PStakeValidator)
+smartHandleStakeValidatorW = phoistAcyclic $ plam $ \validateFn routeAddress redeemer ctx -> P.do
   let red = pconvertUnsafe @PRouterRedeemer redeemer
-  redF <- pletFields @'["inputIdxs", "outputIdxs"] red
+  redF <- pletFields @'["inputIdxs", "outputIdxs", "advancedRedeemers"] red
```
Comment (mpetruska) [inline src/StakingValidator.hs]:
```suggestion
  redF <- pletFields @'["inputIdxs", "outputIdxs"] red
```

File: src/StakingValidator.hs [hunk 4]
```diff
@@ -136,18 +169,61 @@ puniqueOrdered =
...
   ctxF <- pletFields @'["txInfo", "purpose"] ctx
-  infoF <- pletFields @'["inputs", "outputs", "signatories", "datums"] ctxF.txInfo
+  PRewarding _ <- pmatch ctxF.purpose
+  infoF <- pletFields @'["inputs", "outputs", "signatories", "redeemers", "datums"] ctxF.txInfo
```
Comment (mpetruska) [inline src/StakingValidator.hs]:
```suggestion
  infoF <- pletFields @'["inputs", "outputs", "redeemers", "datums"] ctxF.txInfo
```

**Review by mpetruska — CHANGES_REQUESTED** [review]
Nice job, only minor changes requested.

File: src/StakingValidator.hs (hunk 1 above)
Comment (keyan-m) [reply src/StakingValidator.hs:57]: It's for future, to unify the interface with our Aiken's. I believe having a single list of pairs signals the underlying logic better. Although I'm not sure if it'd impact the performance negatively.

If having two lists is cheaper, perhaps we should refine the Aiken's interface?

File: src/BatchValidator.hs [hunk 5]
```diff
@@ -38,23 +37,24 @@ instance PTryFrom PData PSmartRedeemer
 instance PUnsafeLiftDecl PSmartRedeemer where type PLifted PSmartRedeemer = SmartRedeemer
 deriving via (DerivePConstantViaData SmartRedeemer PSmartRedeemer) instance PConstantDecl SmartRedeemer
 
+pstakeScriptIsInvoked :: Term s (PScriptContext :--> PStakingCredential :--> POpaque)
+pstakeScriptIsInvoked = plam $ \ctx stakeScript -> P.do
```
Comment (mpetruska) [inline src/BatchValidator.hs]: How about `passertStakeScriptIsInvoked`?

File: src/Utils.hs [hunk 6]
```diff
@@ -145,3 +204,125 @@ pconvertChecked x = ptryFrom x fst
...
+pgetSingleAssetApartFromADA ::
+  forall
+    (anyOrder :: KeyGuarantees)
+    (anyAmount :: AmountGuarantees)
+    (s :: S).
+  Term s (PValue anyOrder anyAmount) ->
+  Term s (PTriple PCurrencySymbol PTokenName PInteger)
+pgetSingleAssetApartFromADA v =
+  -- not using `pelimList` as it would've lead to evaluation of the head
+  psingleAssetToTriple #$ pheadSingleton $ ptail # presolveValueToList v
```
Comment (mpetruska) [inline src/Utils.hs:293]: This only works for `'Sorted` key guarantee!

**Review by mpetruska — CHANGES_REQUESTED** [review]
Very nice work! I found only one issue.

File: src/Utils.hs (hunk 6 above)
Comment (keyan-m) [reply src/Utils.hs:293]: Very nice catch! Thank you :pray:

**Pull request merged**